Dockerfile

```
@@ -0,0 +1,118 @@
+# syntax = docker/dockerfile:1.4
+
+ARG NVMEOF_SPDK_VERSION
+ARG NVMEOF_TARGET  # either 'gateway' or 'cli'
+
+#------------------------------------------------------------------------------
+# Base image for NVMEOF_TARGET=cli (nvmeof-cli)
+FROM registry.access.redhat.com/ubi9/ubi AS base-cli
+ENTRYPOINT ["python3", "-m", "control.cli"]
+CMD []
+
+#------------------------------------------------------------------------------
+# Base image for NVMEOF_TARGET=gateway (nvmeof-gateway)
+FROM quay.io/ceph/spdk:${NVMEOF_SPDK_VERSION:-NULL} AS base-gateway
+RUN \
+    --mount=type=cache,target=/var/cache/dnf \
+    --mount=type=cache,target=/var/lib/dnf \
+    dnf install -y python3-rados
+ENTRYPOINT ["python3", "-m", "control"]
+CMD ["-c", "/src/ceph-nvmeof.conf"]
+
+#------------------------------------------------------------------------------
+# Intermediate layer for Python set-up
+FROM base-$NVMEOF_TARGET AS python-intermediate
+
+RUN \
+    --mount=type=cache,target=/var/cache/dnf \
+    --mount=type=cache,target=/var/lib/dnf \
+    dnf update -y
+
+ENV PYTHONUNBUFFERED=1 \
+    PYTHONIOENCODING=UTF-8 \
+    LC_ALL=C.UTF-8 \
+    LANG=C.UTF-8 \
+    PIP_NO_CACHE_DIR=off \
+    PYTHON_MAJOR=3 \
+    PYTHON_MINOR=9 \
+    PDM_ONLY_BINARY=:all:
+
+ARG APPDIR=/src
+
+ARG NVMEOF_NAME \
+    NVMEOF_SUMMARY \
+    NVMEOF_DESCRIPTION \
+    NVMEOF_URL \
+    NVMEOF_VERSION \
+    NVMEOF_MAINTAINER \
+    NVMEOF_TAGS \
+    NVMEOF_WANTS \
+    NVMEOF_EXPOSE_SERVICES \
+    BUILD_DATE \
+    NVMEOF_GIT_REPO \
+    NVMEOF_GIT_BRANCH \
+    NVMEOF_GIT_COMMIT
+
+# Generic labels
+LABEL name="$NVMEOF_NAME" \
+      version="$NVMEOF_VERSION" \
+      summary="$NVMEOF_SUMMARY" \
+      description="$NVMEOF_DESCRIPTION" \
+      maintainer="$NVMEOF_MAINTAINER" \
+      release="" \
+      url="$NVMEOF_URL" \
+      build-date="$BUILD_DATE" \
+      vcs-ref="$NVMEOF_GIT_COMMIT"
+
+# k8s-specific labels
+LABEL io.k8s.display-name="$NVMEOF_SUMMARY" \
+      io.k8s.description="$NVMEOF_DESCRIPTION"
+
+# k8s-specific labels
+LABEL io.openshift.tags="$NVMEOF_TAGS" \
+      io.openshift.wants="$NVMEOF_WANTS" \
+      io.openshift.expose-services="$NVMEOF_EXPOSE_SERVICES"
+
+# Ceph-specific labels
+LABEL io.ceph.component="$NVMEOF_NAME" \
+      io.ceph.summary="$NVMEOF_SUMMARY" \
+      io.ceph.description="$NVMEOF_DESCRIPTION" \
+      io.ceph.url="$NVMEOF_URL" \
+      io.ceph.version="$NVMEOF_VERSION" \
+      io.ceph.maintainer="$NVMEOF_MAINTAINER" \
+      io.ceph.git.repo="$NVMEOF_GIT_REPO" \
+      io.ceph.git.branch="$NVMEOF_GIT_BRANCH" \
+      io.ceph.git.commit="$NVMEOF_GIT_COMMIT"
+
```

```suggestion
RUN echo echo $NVMEOF_NAME $NVMEOF_VERSION $NVMEOF_GIT_COMMIT > /bin/ceph
RUN chmod 755 /bin/ceph
```

cephadm compatibility

Add `ceph` executable printing this component version, required by cephadm

cephadm `command_pull()`, runs `command_inspect_image()` which execs in container "ceph --version", see https://github.com/rkachach/ceph/blob/d6dfea39387722ca688cfa0bdfd4f16bfa0ec3cb/src/cephadm/cephadm.py#L5065-L5066
    
Example error:
```shell
RuntimeError: Failed command: /usr/bin/podman run --rm --ipc=host --stop-signal=SIGTERM --net=host --entrypoint ceph --init -e CONTAINER_IMAGE=quay.io/ceph/nvmeof:0.0.1 -e NODE_NAME=centos9 quay.io/ceph/nvmeof:0.0.1 --version: ERROR (catatonit:2): failed to exec pid1: No such file or directory
```

---

Thanks @baum!

Rather than creating a fake `ceph` binary, I'd add a `--version` flag support to the current entrypoint (`python3 -m control`), so that cephadm doesn't have to override the `entrypoint`:
```diff
- /usr/bin/podman run --rm --ipc=host --stop-signal=SIGTERM --net=host --entrypoint ceph --init -e CONTAINER_IMAGE=quay.io/ceph/nvmeof:0.0.1 -e NODE_NAME=centos9 quay.io/ceph/nvmeof:0.0.1 --version
+ /usr/bin/podman run --rm --ipc=host --stop-signal=SIGTERM --net=host --init -e CONTAINER_IMAGE=quay.io/ceph/nvmeof:0.0.1 -e NODE_NAME=centos9 quay.io/ceph/nvmeof:0.0.1 --version
```
@rkachach what do you think? Another approach?

---

> Thanks @baum!
> 
> Rather than creating a fake `ceph` binary, I'd add a `--version` flag support to the current entrypoint (`python3 -m control`), so that cephadm doesn't have to override the `entrypoint`:

@epuertat this is a common image pull path in cephadm, unrelated to nvmeof, the entry point is hardcoded as `ceph` see [here](https://github.com/ceph/ceph/blob/fd62a8dfcb345fabed3eea6b3a461f1ca848a239/src/cephadm/cephadm.py#L5007):

```python
    ver = CephContainer(ctx, ctx.image, 'ceph', ['--version']).run().strip()
```

---

As @baum pointed out this is a generic path in cephadm code and the boostrap process assumes that the image used is a `ceph based image` (so it must include the ceph binary).

---

@baum @rkachach as per our conversation I understand this wouldn't be required, right?

---

> @baum @rkachach as per our conversation I understand this wouldn't be required, right?

@epuertat 🖖,  The 'ceph' executable is necessary within the container image because the cephadm container management relies on its existence.

---

@baum, @rkachach told me otherwise, that this is only required for the bootstrap phase. If that were the case, we couldn't deploy Grafana, Prometheus or any other non-Ceph container...

---

@epuertat 🙂 running "ceph --version" is a step of the pull image procedure.  In a cephadm container management environment, if the 'ceph' executable is not present in the deployed image, that deployment would fail with a `RuntimeError: Failed command`.   Therefore 'ceph' executable is required for images targeted to be deployed by cephadm.

---

@rkachach 🖖, could you share your thoughts?

---

@epuertat @baum as I understood from @baum feedback the error was coming from the method:

```
    @staticmethod
    def get_version(ctx, container_id):
```

This method is meant to get the version of the container itself. In the case of iscsi for example it uses the version of the rpm package:
```
 '/usr/bin/python3 -c "import pkg_resources; print(pkg_resources.require('ceph_iscsi')[0].version)"
```

Not sure what would be the equivalent for nvmeof (nor if we will have an rpm package for the same or not).

As for the other issue, in principle ceph binary is not needed for daemons. It's called during bootstrap because this is normally run with the ceph image (which contains the binary obviously).

---

Good thing of the approach proposed by @baum (inspecting labels) is that it doesn't require to run the container, so it should be extremely efficient. If we wanted to keep consistency with other Ceph Python containers, we could do something like:
```
/usr/bin/python3 -c "print(__import__('control').__version__)""
```
But then we need to restore the `__version__ = x.y.z` symbol to the `__init__.py`. Given that we don't plan to release the NVMe-oF GW in other format than a container, I think that using `LABELs` and/or `ENV` variables is probably the best way.

---

BTW, our intention is that this `io.ceph.version` matches the container image tag (`nvmeof:<io.ceph.version>`), which will be hard-coded in the `cephadm` source code, right? Would still be necessary to dynamically discover that?

---

@epuertat awesome, it would be great if you provide me with a simple python code to get the version I use it as cephadm `def get_version()` method.